src/core/PRINCIPLES.md

```diff
@@ -0,0 +1,47 @@
+## Common dictionary
+Plugin - a piece of software expressing specific subject of the business unit within the code.
+
+Platform - a set of plugins required to run all other Kibana plugins.
+
+## New platform principles
+### Explicit business domains separation
+The plugins code should be structured on the top level to reflect business units.
+```js
+// GOOD
+src/plugins
+ - apm
+  - server
+  - public
+ - search
+ ..
+// BAD
+src/plugins
+ - server
+  - apm
+  - search
+```
+### Explicit dependencies
+Each plugin should declare dependencies on the other plugins explicitly. Plugins cannot have circular dependencies. Plugins shouldn't have hidden dependencies.
```

Comment:
Plugins should also not have hard dependencies on registered values. For example, one plugin should handle the case where a specific visualization type has not been registered, unless it has declared an explicit dependency on that plugin.

Reply:
I see the point. This is a particular case of `Plugins shouldn't have hidden dependencies.` Don't know if we should provide a specific example of it. Otherwise, the file will grow to an unmanageable size quite soon. WDYT?

Reply:
I think this is a pretty common one that I've had to warn about before. Though TypeScript does help us out here now, I'm not sure all teams are going to be using TS everywhere.

Reply:
what if express it as `Plugins shouldn't access runtime objects, HTTP endpoints, DOM nodes, etc. created by a third party plugin without declaring a dependency on this plugin.`